src/nbl/ext/ImGui/ImGui.cpp

```diff
@@ -3,957 +3,1496 @@
-	{
-		ImGuiIO& io = ImGui::GetIO();
-		// TODO:
-		// Keyboard mapping. ImGui will use those indices to peek into the io.KeysDown[] array.
-		//io.KeyMap[ImGuiKey_Tab] = MSDL::SDL_SCANCODE_TAB;
-		//io.KeyMap[ImGuiKey_LeftArrow] = MSDL::SDL_SCANCODE_LEFT;
-		//io.KeyMap[ImGuiKey_RightArrow] = MSDL::SDL_SCANCODE_RIGHT;
-		//io.KeyMap[ImGuiKey_UpArrow] = MSDL::SDL_SCANCODE_UP;
-		//io.KeyMap[ImGuiKey_DownArrow] = MSDL::SDL_SCANCODE_DOWN;
-		//io.KeyMap[ImGuiKey_PageUp] = MSDL::SDL_SCANCODE_PAGEUP;
-		//io.KeyMap[ImGuiKey_PageDown] = MSDL::SDL_SCANCODE_PAGEDOWN;
-		//io.KeyMap[ImGuiKey_Home] = MSDL::SDL_SCANCODE_HOME;
-		//io.KeyMap[ImGuiKey_End] = MSDL::SDL_SCANCODE_END;
-		//io.KeyMap[ImGuiKey_Insert] = MSDL::SDL_SCANCODE_INSERT;
-		//io.KeyMap[ImGuiKey_Delete] = MSDL::SDL_SCANCODE_DELETE;
-		//io.KeyMap[ImGuiKey_Backspace] = MSDL::SDL_SCANCODE_BACKSPACE;
-		//io.KeyMap[ImGuiKey_Space] = MSDL::SDL_SCANCODE_SPACE;
-		//io.KeyMap[ImGuiKey_Enter] = MSDL::SDL_SCANCODE_RETURN;
-		//io.KeyMap[ImGuiKey_Escape] = MSDL::SDL_SCANCODE_ESCAPE;
-		//io.KeyMap[ImGuiKey_KeyPadEnter] = MSDL::SDL_SCANCODE_KP_ENTER;
-		//io.KeyMap[ImGuiKey_A] = MSDL::SDL_SCANCODE_A;
-		//io.KeyMap[ImGuiKey_C] = MSDL::SDL_SCANCODE_C;
-		//io.KeyMap[ImGuiKey_V] = MSDL::SDL_SCANCODE_V;
-		//io.KeyMap[ImGuiKey_X] = MSDL::SDL_SCANCODE_X;
-		//io.KeyMap[ImGuiKey_Y] = MSDL::SDL_SCANCODE_Y;
-		//io.KeyMap[ImGuiKey_Z] = MSDL::SDL_SCANCODE_Z;
-	}
-
-	static void adjustGlobalFontScale()
-	{
-		ImGuiIO& io = ImGui::GetIO();
-		io.FontGlobalScale = 1.0f;
-	}
-
-	void UI::UpdateDescriptorSets()
-	{
-		IGPUDescriptorSet::SDescriptorInfo info;
+			return gpu;
+		};
+
+		if (!system->areBuiltinsMounted())
 		{
-			info.desc = m_fontTexture;
-			info.info.combinedImageSampler.sampler = m_fontSampler;
-			info.info.combinedImageSampler.imageLayout = nbl::asset::IImage::LAYOUT::READ_ONLY_OPTIMAL;
+			creationParams.utilities->getLogger()->log("Nabla builtins are not mounted!", ILogger::ELL_ERROR);
+			return nullptr;
 		}
 
-		IGPUDescriptorSet::SWriteDescriptorSet writeDescriptorSet{};
-		writeDescriptorSet.dstSet = m_gpuDescriptorSet.get();
-		writeDescriptorSet.binding = 0;
-		writeDescriptorSet.arrayElement = 0;
-		writeDescriptorSet.count = 1;
-		writeDescriptorSet.info = &info;
+		//! but we should never assume user will mount our internal archive since its the extension and not user's job to do it, hence we mount ourselves temporary archive to compile our extension sources then unmount it
+		auto archive = mount(smart_refctd_ptr<ILogger>(creationParams.utilities->getLogger()), system.get(), NBL_ARCHIVE_ALIAS.data());
+		shaders.vertex = createShader.template operator() < NBL_CORE_UNIQUE_STRING_LITERAL_TYPE("vertex.hlsl"), IShader::E_SHADER_STAGE::ESS_VERTEX > ();
+		shaders.fragment = createShader.template operator() < NBL_CORE_UNIQUE_STRING_LITERAL_TYPE("fragment.hlsl"), IShader::E_SHADER_STAGE::ESS_FRAGMENT > ();
+		system->unmount(archive.get(), NBL_ARCHIVE_ALIAS.data());
```

Comment:
as per the discord conversation:
1. never unmount (user will do if they want)
2. only call `mount(` if there's no archive mounted already at `NBL_ARCHIVE_ALIAS`

Reply:
resolved